resolwe/flow/migration_ops.py

```diff
@@ -0,0 +1,373 @@
+        raise NotImplementedError
+
+
+class ResolweProcessAddField(ResolweProcessOperation):  # pylint: disable=abstract-method
+    """Operation, which adds a new field to existing process schema."""
+
+    def __init__(self, process, field, schema, default=None):
+        """Construct a process field add operation.
+
+        :param process: Process slug
+        :param field: Dot-separated field path, starting with the top-level
+            database field name (must be either ``input`` or ``output``)
+        :param schema: Field schema definition
+        :param default: An operation that generates default values, which
+            must be an instance of :class:`DataDefaultOperation` or its
+            subclass
+        """
+        self._raw_field = field
+        self.field = field.split('.')
+        if self.field[0] not in ('input', 'output'):
+            raise ValueError("Field path must start with either input or output")
+        if len(self.field) < 2:
+            raise ValueError("Field path must contain at least two levels")
+        if len(self.field) != 2:
+            # TODO: Support nested fields.
+            raise NotImplementedError("Field path must contain exactly two levels")
+        schema_type = self.field.pop(0)
+
+        self.schema = schema
+        jsonschema.validate([schema], FIELD_SCHEMA)
+
+        if schema['name'] != self.field[-1]:
+            raise ValueError("Field name in schema differs from field path")
+
+        if not default and schema.get('required', True):
+            raise ValueError("A required field must have a default")
+        if default and not isinstance(default, DataDefaultOperation):
+            raise TypeError("Default must be an instance of DataDefaultOperation")
+        self.default = default
+
+        super(ResolweProcessAddField, self).__init__(
+            process=process,
+            schema_type=schema_type,
+            field=field,
+            schema=schema,
+            default=default
+        )
+
+    def deconstruct(self):
+        """Deconstruct operation."""
+        return (
+            self.__class__.__name__,
+            [],
+            {
+                'process': self.process,
+                'field': self._raw_field,
+                'schema': self.schema,
+                'default': self.default,
+            }
+        )
+
+    def migrate_process_schema(self, process, schema, from_state):
+        """Migrate process schema.
+
+        :param process: Process instance
+        :param schema: Process schema to migrate
+        :param from_state: Database model state
+        :return: True if the process was migrated, False otherwise
+        """
+        container = dict_dot(schema, '.'.join(self.field[:-1]))
+        assert container
+
+        # Ignore processes, which already contain the target field.
+        for field in container:
+            if field['name'] == self.field[-1]:
+                return False
```

Comment:
If field already exists, this will cause an inconsistent state. Maybe it would be better to raise an error?

Reply:
:+1:

Reply:
Oh, now I remember what the issue is here. There can be multiple process versions and the question is which ones to migrate? The latest version may already contain the given field, so it should probably be skipped? This is why it is currently made to skip by default.

How should we handle this? Error on all except latest if they already have the field and error on latest if it doesn't have the field?

Reply:
Can you raise error only if schemas don't match?